src/PackageManagement/NuGetPackageManager.cs

```diff
@@ -1638,27 +1638,62 @@ private Task OpenReadmeFile(INuGetProjectContext nuGetProjectContext, Cancellati
             }
 
             token.ThrowIfCancellationRequested();
-            nuGetProjectContext.Log(ProjectManagement.MessageLevel.Info, string.Format(Strings.RestoringPackage, packageIdentity));
+
+            nuGetProjectContext.Log(
+                ProjectManagement.MessageLevel.Info,
+                string.Format(Strings.RestoringPackage, packageIdentity));
+
             var enabledSources = (sourceRepositories != null && sourceRepositories.Any()) ? sourceRepositories :
                 SourceRepositoryProvider.GetRepositories().Where(e => e.PackageSource.IsEnabled);
 
             token.ThrowIfCancellationRequested();
 
-            using (var downloadResult = await PackageDownloader.GetDownloadResourceResultAsync(enabledSources,
-                packageIdentity,
-                Settings,
-                token))
+            // maximum number of times to try restoring the package
+            var maxTries = 3;
+            var success = false;
+
+            for (int i = 0; i < maxTries && !token.IsCancellationRequested && !success; i++)
             {
-                packageIdentity = downloadResult.PackageReader.GetIdentity();
+                try
+                {
+                    using (var downloadResult = await PackageDownloader.GetDownloadResourceResultAsync(
+                        enabledSources,
+                        packageIdentity,
+                        Settings,
+                        token))
+                    {
+                        packageIdentity = downloadResult.PackageReader.GetIdentity();
+
+                        // If you already downloaded the package, just restore it, don't cancel the operation now
+                        await PackagesFolderNuGetProject.InstallPackageAsync(
+                            packageIdentity,
+                            downloadResult,
+                            nuGetProjectContext,
+                            token);
 
-                // If you already downloaded the package, just restore it, don't cancel the operation now
-                await PackagesFolderNuGetProject.InstallPackageAsync(packageIdentity, downloadResult, nuGetProjectContext, token);
+                        success = true;
+                    }
+                }
+                catch (Exception) when (i < (maxTries - 1))
```

Comment:
Those exceptions are already retried. This is just an additional retry since for the majority of packages.config restore scenarios the packages are expected to exist, if they cannot be found due to an error we just retry an extra time to save the user the hassle of attempting the retry again themselves.

To fix this further it would need to be fixed in NuGet.Core on the DataServicePackageRepository. These changes are too risky for 3.2.

I will remove this change and disable the test this change was made to fix.